graylog2-server/src/main/java/org/graylog2/rest/resources/datanodes/DatanodeRestApiProxy.java

```diff
@@ -32,37 +36,76 @@
 import org.graylog2.security.IndexerJwtAuthTokenProvider;
 import org.jetbrains.annotations.NotNull;
 
+import java.io.ByteArrayInputStream;
 import java.io.IOException;
-import java.security.KeyManagementException;
-import java.security.NoSuchAlgorithmException;
-import java.util.Collection;
+import java.io.InputStream;
+import java.time.Duration;
 import java.util.Locale;
+import java.util.Map;
 import java.util.Objects;
-import java.util.Optional;
+import java.util.stream.Collectors;
 
 @Singleton
 public class DatanodeRestApiProxy implements ProxyRequestAdapter {
 
     private final IndexerJwtAuthTokenProvider authTokenProvider;
+    private final NodeService<DataNodeDto> nodeService;
+    private final ObjectMapper objectMapper;
     private final DatanodeResolver datanodeResolver;
     private final OkHttpClient httpClient;
 
     @Inject
-    public DatanodeRestApiProxy(IndexerJwtAuthTokenProvider authTokenProvider, DatanodeResolver datanodeResolver, OkHttpClient okHttpClient) throws NoSuchAlgorithmException, KeyManagementException {
+    public DatanodeRestApiProxy(IndexerJwtAuthTokenProvider authTokenProvider, NodeService<DataNodeDto> nodeService, ObjectMapper objectMapper, DatanodeResolver datanodeResolver, OkHttpClient okHttpClient, @Named("proxied_requests_default_call_timeout")
+    com.github.joschi.jadconfig.util.Duration defaultProxyTimeout) {
         this.authTokenProvider = authTokenProvider;
+        this.nodeService = nodeService;
+        this.objectMapper = objectMapper;
         this.datanodeResolver = datanodeResolver;
-        httpClient = okHttpClient;
+        httpClient = withTimeout(okHttpClient, defaultProxyTimeout);
+    }
+
+    @NotNull
+    private static OkHttpClient withTimeout(OkHttpClient okHttpClient, com.github.joschi.jadconfig.util.Duration defaultProxyTimeout) {
+        final Duration timeout = Duration.ofMillis(defaultProxyTimeout.toMilliseconds());
+        return okHttpClient.newBuilder()
+                .connectTimeout(timeout)
+                .readTimeout(timeout)
+                .callTimeout(timeout)
+                .build();
+    }
+
+    private ProxyResponse runOnAllNodes(ProxyRequest request) {
+        final Map<String, JsonNode> result = nodeService.allActive().values().stream().parallel().collect(Collectors.toMap(NodeDto::getHostname, n -> {
+            try {
+                final ProxyResponse response = request(new ProxyRequest(request.method(), request.path(), InputStream.nullInputStream(), n.getHostname()));
+                return objectMapper.readValue(response.response(), JsonNode.class);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }));
+
+        try {
+            ByteArrayInputStream bais = new ByteArrayInputStream(objectMapper.writeValueAsBytes(result));
+            return new ProxyResponse(200, bais, "application/json");
```

Comment:
Maybe find/create a `Ok` constant instead of the `200`?

Reply:
Good point, fixed